core/src/consensus/fork_choice.rs

```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
+    // reset to the the next votable bank on the same fork as our last vote,
+    // but don't vote.
+
+    // We don't just reset to the heaviest fork when switch threshold fails because
+    // a situation like this can occur:
+
+    /* Figure 1:
+                slot 0
+                    |
+                slot 1
+                /        \
+    slot 2 (last vote)     |
+                |      slot 8 (10%)
+        slot 4 (9%)
+    */
+
+    // Imagine 90% of validators voted on slot 4, but only 9% landed. If everybody that fails
+    // the switch threshold abandons slot 4 to build on slot 8 (because it's *currently* heavier),
+    // then there will be no blocks to include the votes for slot 4, and the network halts
+    // because 90% of validators can't vote
+    info!(
+        "Waiting to switch vote to {},
+        resetting to slot {:?} for now,
+        switch proof stake: {},
+        threshold stake: {},
+        total stake: {}",
+        heaviest_bank_slot,
+        reset_bank.as_ref().map(|b| b.slot()),
+        switch_proof_stake,
+        total_stake as f64 * SWITCH_FORK_THRESHOLD,
+        total_stake
+    );
+    failure_reasons.push(HeaviestForkFailures::FailedSwitchThreshold(
+        heaviest_bank_slot,
+        switch_proof_stake,
+        total_stake,
+    ));
+    switch_fork_decision
+}
+
+fn select_candidate_vote_and_reset_banks<'a>(
+    heaviest_bank: &'a Arc<Bank>,
+    heaviest_bank_on_same_voted_fork: Option<&'a Arc<Bank>>,
+    progress: &'a ProgressMap,
+    tower: &'a Tower,
+    failure_reasons: &mut Vec<HeaviestForkFailures>,
+    initial_switch_fork_decision: SwitchForkDecision,
+) -> CandidateVoteAndResetBanks<'a> {
+    match initial_switch_fork_decision {
+        SwitchForkDecision::FailedSwitchThreshold(switch_proof_stake, total_stake) => {
+            // Even though we think we want to switch to a new, heavier fork, we
+            // actually want to continue voting on our current fork if our last
+            // vote was unable land.
+            let final_switch_fork_decision = select_forks_failed_switch_threshold(
+                heaviest_bank_on_same_voted_fork.map(|bank| bank.as_ref()),
+                progress,
+                tower,
+                heaviest_bank.slot(),
+                failure_reasons,
+                switch_proof_stake,
+                total_stake,
+                initial_switch_fork_decision,
+            );
+            let candidate_vote_bank = if final_switch_fork_decision.can_vote() {
+                heaviest_bank_on_same_voted_fork
+            } else {
+                Some(heaviest_bank)
+            };
+            CandidateVoteAndResetBanks {
+                candidate_vote_bank,
+                reset_bank: heaviest_bank_on_same_voted_fork,
+                switch_fork_decision: final_switch_fork_decision,
+            }
+        }
+        SwitchForkDecision::FailedSwitchDuplicateRollback(latest_duplicate_ancestor) => {
```

Comment:
Added back in and tweaked the wording slightly.

This also triggered me to create sub-functions for each leg of the match because the comments are so giant that the `select_candidate_vote_and_reset_banks` function becomes huge.